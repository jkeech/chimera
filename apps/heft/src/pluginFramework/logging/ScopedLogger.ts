// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Terminal, ITerminalProvider } from '@rushstack/node-core-library';

import { IHeftPlugin } from '../IHeftPlugin';
import { PrefixProxyTerminalProvider } from '../../utilities/PrefixProxyTerminalProvider';
import { LoggingManager } from './LoggingManager';

export interface IScopedLoggerOptions {
  requestingPlugin: IHeftPlugin;
  loggerName: string;
  terminalProvider: ITerminalProvider;
  getShouldPrintStacks: () => boolean;
  errorHasBeenEmittedCallback: () => void;
}

/**
 * @public
 */
export interface IScopedLogger {
  readonly terminal: Terminal;

  /**
   * Call this function to emit an error to the heft runtime.
   */
  emitError(error: Error): void;

  /**
   * Call this function to emit an warning to the heft runtime.
   */
  emitWarning(warning: Error): void;
}

/**
 * @public
 */
export class ScopedLogger implements IScopedLogger {
  private readonly _options: IScopedLoggerOptions;
  private readonly _errors: Error[] = [];
  private readonly _warnings: Error[] = [];

  private get _shouldPrintStacks(): boolean {
    return this._options.getShouldPrintStacks();
  }

  public get errors(): ReadonlyArray<Error> {
    return [...this._errors];
  }

  public get warnings(): ReadonlyArray<Error> {
    return [...this._warnings];
  }

  /**
   * @internal
   */
  public readonly _requestingPlugin: IHeftPlugin;

  public readonly loggerName: string;

  public readonly terminalProvider: ITerminalProvider;

  public readonly terminal: Terminal;

  /**
   * @internal
   */
  public constructor(options: IScopedLoggerOptions) {
    this._options = options;
    this._requestingPlugin = options.requestingPlugin;
    this.loggerName = options.loggerName;

    this.terminalProvider = new PrefixProxyTerminalProvider(
      options.terminalProvider,
      `[${this.loggerName}] `
    );
    this.terminal = new Terminal(this.terminalProvider);
  }

  /**
   * {@inheritdoc IScopedLogger.emitError}
   */
  public emitError(error: Error): void {
    this._errors.push(error);
    this.terminal.writeErrorLine(`Error: ${LoggingManager.getErrorMessage(error)}`);
    if (this._shouldPrintStacks && error.stack) {
      this.terminal.writeErrorLine(error.stack);
    }
  }

  /**
   * {@inheritdoc IScopedLogger.emitWarning}
   */
  public emitWarning(warning: Error): void {
    this._warnings.push(warning);
    this.terminal.writeWarningLine(`Warning: ${LoggingManager.getErrorMessage(warning)}`);
    if (this._shouldPrintStacks && warning.stack) {
      this.terminal.writeWarningLine(warning.stack);
    }
  }
}
