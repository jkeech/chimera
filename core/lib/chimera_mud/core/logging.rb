# frozen_string_literal: true

require "active_support/concern"

module ChimeraMud
  module Core
    ##
    # Adds the logger method. Useful for things that use RactorCommon
    module Logging
      extend ActiveSupport::Concern

      def logger
        @logger ||= Logger.new($stdout)
      end
    end
  end
end
