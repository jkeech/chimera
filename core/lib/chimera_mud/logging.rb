# frozen_string_literal: true

require "chimera_mud/logging/tagged_logger"

module ChimeraMud
  ##
  # This module provids logging functionality for ChimeraMud components.
  module Logging
  end
end
