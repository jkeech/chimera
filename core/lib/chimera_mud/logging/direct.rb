# frozen_string_literal: true

require "chimera/logging/formatter"

module ChimeraMud
  module Logging
    ##
    # Adds logging capabilities to whatever class this module is incuded. The
    # log file and level are automatically determined by the configuration of
    # the game
    module Direct
      extend ActiveSupport::Concern

      # :nodoc:
      module ClassMethods
        ##
        # Sets the tag that the logger for this class will use
        #
        # @param tag [String] the tag to use for the log
        #
        # @return [String] the tag set for this class
        def log_tag(tag = nil)
          @log_tag = tag if tag
          @log_tag
        end

        ##
        # @return [TaggedLogger] the log instance for the class. Note: the log
        # instance for the class will be different than the log instance for
        # the instance of the class
        def logger
          @logger ||= TaggedLogger.new(log_tag)
        end
      end
    end
  end
end
