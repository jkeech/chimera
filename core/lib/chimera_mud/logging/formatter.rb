# frozen_string_literal: true

module ChimeraMud
  module Logging
    ##
    # Formatter for ChimeraMud logging.
    class Formatter < Logger::Formatter
      ##
      # @param [String] tag the tag to append to log lines.
      def initialize(tag = nil)
        super()
        @tag = tag
      end

      def call(severity, time, _progname, msg)
        return "#{severity}, [#{time} ##{$$}] - <#{tag}}> -- #{msg}" if tag

        "#{severity}, [#{time} ##{$$}] -- #{msg}"
      end
    end
  end
end
