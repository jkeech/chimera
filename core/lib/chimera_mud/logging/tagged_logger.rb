# frozen_string_literal: true

require "chimera_mud/logging/formatter"

module ChimeraMud
  module Logging
    ##
    # The tagged logger provides logging functionality and formatting for
    # ChimeraMud, and attaches "tags" to help developers and admins to
    # determin what processes or threads are doing the logging
    class TaggedLogger < Logger
      ##
      # @param [String] tag the tag to append to all lines for this logger
      # instance
      def initialize(tag)
        super(log_file)
        @tag = tag
        self.formatter = Formatter.new(tag)
      end

      private

      def log_file
        ENV.fetch("CHIMERA_LOG_FILE", "/dev/stdout")
      end

      def level
        @level ||= ENV.fetch(
          "CHIMERA_LOG_LEVEL",
          (Rails.env.production? ? "info" : "debug")
        )
      end
    end
  end
end
