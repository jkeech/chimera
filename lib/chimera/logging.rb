# frozen_string_literal: true

require "active_support/concern"

module Chimera
  ##
  # Adds the logger method. Useful for things that use RactorCommon
  module Logging
    extend ActiveSupport::Concern

    def logger
      Rails.logger
    end
  end
end
