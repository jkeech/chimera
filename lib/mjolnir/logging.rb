# frozen_string_literal: true

require "mjolnir/logging/direct"
require "mjolnir/logging/tagged_logger"

module Mjolnir
  ##
  # This module provids logging functionality for Mjolnir components.
  module Logging
  end
end
