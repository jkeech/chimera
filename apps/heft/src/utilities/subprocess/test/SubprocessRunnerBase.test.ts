// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { SubprocessRunnerBase } from '../SubprocessRunnerBase';
import { FileError } from '../../../pluginFramework/logging/FileError';

describe('SubprocessRunnerBase', () => {
  it(`${SubprocessRunnerBase.serializeForIpcMessage.name} correctly serializes objects`, () => {
    expect(SubprocessRunnerBase.serializeForIpcMessage(1)).toMatchSnapshot();
    expect(SubprocessRunnerBase.serializeForIpcMessage(false)).toMatchSnapshot();
    expect(SubprocessRunnerBase.serializeForIpcMessage('abc')).toMatchSnapshot();
    expect(SubprocessRunnerBase.serializeForIpcMessage(null)).toMatchSnapshot();
    expect(SubprocessRunnerBase.serializeForIpcMessage(undefined)).toMatchSnapshot();
    const error: Error = new Error();
    error.stack = 'ERROR STACK';
    expect(SubprocessRunnerBase.serializeForIpcMessage(error)).toMatchSnapshot();
    const fileError1: FileError = new FileError('message', 'path/to/file', 4, 29);
    fileError1.stack = 'ERROR STACK';
    expect(SubprocessRunnerBase.serializeForIpcMessage(fileError1)).toMatchSnapshot();
    const fileError2: FileError = new FileError('message', 'path/to/file', 4);
    fileError2.stack = 'ERROR STACK';
    expect(SubprocessRunnerBase.serializeForIpcMessage(fileError2)).toMatchSnapshot();
    const fileError3: FileError = new FileError('message', 'path/to/file');
    fileError3.stack = 'ERROR STACK';
    expect(SubprocessRunnerBase.serializeForIpcMessage(fileError3)).toMatchSnapshot();
  });

  it(`${SubprocessRunnerBase.serializeForIpcMessage.name} doesn't handle non-error objects`, () => {
    expect(() => SubprocessRunnerBase.serializeForIpcMessage({})).toThrow();
  });

  it('de-serializes serialized objects', () => {
    function testDeserialization(x: unknown): void {
      expect(
        SubprocessRunnerBase.deserializeFromIpcMessage(SubprocessRunnerBase.serializeForIpcMessage(x))
      ).toEqual(x);
    }

    testDeserialization(1);
    testDeserialization(false);
    testDeserialization('abc');
    testDeserialization(null);
    testDeserialization(undefined);
    testDeserialization(new Error());
    const fileError1: FileError = new FileError('message', 'path/to/file', 4, 29);
    testDeserialization(fileError1);
    const fileError2: FileError = new FileError('message', 'path/to/file', 4);
    testDeserialization(fileError2);
    const fileError3: FileError = new FileError('message', 'path/to/file');
    testDeserialization(fileError3);
  });
});
