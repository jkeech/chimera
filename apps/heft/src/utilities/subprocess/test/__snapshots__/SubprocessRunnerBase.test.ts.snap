// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubprocessRunnerBase serializeForIpcMessage correctly serializes objects 1`] = `
Object {
  "type": 2,
  "value": 1,
}
`;

exports[`SubprocessRunnerBase serializeForIpcMessage correctly serializes objects 2`] = `
Object {
  "type": 2,
  "value": false,
}
`;

exports[`SubprocessRunnerBase serializeForIpcMessage correctly serializes objects 3`] = `
Object {
  "type": 2,
  "value": "abc",
}
`;

exports[`SubprocessRunnerBase serializeForIpcMessage correctly serializes objects 4`] = `
Object {
  "type": 1,
}
`;

exports[`SubprocessRunnerBase serializeForIpcMessage correctly serializes objects 5`] = `
Object {
  "type": 0,
}
`;

exports[`SubprocessRunnerBase serializeForIpcMessage correctly serializes objects 6`] = `
Object {
  "type": 3,
  "value": Object {
    "errorMessage": "",
    "errorStack": "ERROR STACK",
  },
}
`;

exports[`SubprocessRunnerBase serializeForIpcMessage correctly serializes objects 7`] = `
Object {
  "type": 4,
  "value": Object {
    "column": 29,
    "errorMessage": "message",
    "errorStack": "ERROR STACK",
    "filePath": "path/to/file",
    "line": 4,
  },
}
`;

exports[`SubprocessRunnerBase serializeForIpcMessage correctly serializes objects 8`] = `
Object {
  "type": 4,
  "value": Object {
    "column": undefined,
    "errorMessage": "message",
    "errorStack": "ERROR STACK",
    "filePath": "path/to/file",
    "line": 4,
  },
}
`;

exports[`SubprocessRunnerBase serializeForIpcMessage correctly serializes objects 9`] = `
Object {
  "type": 4,
  "value": Object {
    "column": undefined,
    "errorMessage": "message",
    "errorStack": "ERROR STACK",
    "filePath": "path/to/file",
    "line": undefined,
  },
}
`;
